src/base/projects/Project.ts

```diff
@@ -0,0 +1,338 @@
+  info: EntryInfo;
+}
+
+export interface Blob {
+  binary?: boolean;
+  html?: boolean;
+  content: string;
+  path: string;
+  info: EntryInfo;
+}
+
+export interface Remote {
+  heads: Branches;
+}
+
+export interface Browser {
+  content: ProjectContent;
+  revision: string | null;
+  peer: string | null;
+  path: string | null;
+}
+
+export const browserStore = writable({
+  content: ProjectContent.Tree,
+  branches: {},
+  revision: null,
+  peer: null,
+  path: null,
+} as Browser);
+
+export interface BrowseTo {
+  content?: ProjectContent;
+  revision?: string | null;
+  path?: string | null;
+  peer?: string | null;
+}
+
+export interface PathOptions extends BrowseTo {
+  urn: string;
+  profile?: string | null;
+  seed?: string | null;
+}
+
+export function browse(browse: BrowseTo): void {
+  const browser = get(browserStore);
+  browserStore.set({ ...browser, ...browse });
+}
+
+export function path(opts: PathOptions): string {
+  const { urn, profile, seed, peer, content, revision, path } = opts;
+  const result = [];
+
+  if (profile) {
+    result.push(profile);
+  } else if (seed) {
+    result.push("seeds", seed);
+  }
+  result.push(urn);
+
+  if (peer) {
+    result.push("remotes", peer);
+  }
+
+  switch (content) {
+    case ProjectContent.History:
+      result.push("history");
+      break;
+
+    case ProjectContent.Commit:
+      result.push("commits");
+      break;
+
+    default:
+      result.push("tree");
+      break;
+  }
+
+  if (revision) {
+    result.push(revision);
+  }
+
+  // Avoids appending a slash when the path is the root directory.
+  if (path && path !== "/") {
+    result.push(path);
+  }
+  return "/" + result.join("/");
+}
+
+// We need a SHA1 commit in some places, so we return early if the revision is a SHA and else we look into branches.
+// As fallback we use the head commit.
+export function getOid(revision: string, branches?: Branches): string | null {
+  if (isOid(revision)) return revision;
+
+  if (branches) {
+    const oid = branches[revision];
+
+    if (oid) {
+      return oid;
+    }
+  }
+  return null;
+}
+
+// Parses the path consisting of a revision (eg. branch or commit) and file path into a tuple [revision, file-path]
+export function parseRoute(input: string, branches: Branches): { path?: string; revision?: string } {
+  const branch = Object.entries(branches).find(([branchName,]) => input.startsWith(branchName));
+  const commitPath = [input.slice(0, 40), input.slice(41)];
+  const parsed: { path?: string; revision?: string } = {};
+
+  if (branch) {
+    const [rev, path] = [input.slice(0, branch[0].length), input.slice(branch[0].length + 1)];
+
+    parsed.revision = rev;
+    parsed.path = path ? path : "/";
+  } else if (isOid(commitPath[0])) {
+    parsed.revision = commitPath[0];
+    parsed.path = commitPath[1] ? commitPath[1] : "/";
+  } else {
+    parsed.path = input;
+  }
+  return parsed;
+}
+
+export class Project {
+  urn: string;
+  info: ProjectInfo;
```

Comment:
I would extend `Project` with `ProjectInfo`. Then you can pass `Project` where `ProjectInfo` is expected, eg. BranchSelector.